src/CodeGen_RISCV.cpp

```diff
@@ -1,24 +1,114 @@
+#include "CSE.h"
+#include "CodeGen_Internal.h"
 #include "CodeGen_Posix.h"
+#include "ConciseCasts.h"
+#include "Debug.h"
+#include "IREquality.h"
+#include "IRMatch.h"
+#include "IROperator.h"
+#include "IRPrinter.h"
+#include "LLVM_Headers.h"
+#include "Simplify.h"
+#include "Substitute.h"
+#include "Util.h"
 
 namespace Halide {
 namespace Internal {
 
 using std::string;
+using std::vector;
 
 #if defined(WITH_RISCV)
 
 namespace {
 
+constexpr int max_intrinsic_args = 4;
+
+struct FixedOrScalableVector {
+    enum TypePattern {
+        Undefined,
+        Fixed,
+        Scalable,
+        WildcardWidths,
+    } type_pattern;
+
+    Type type;
+    int relative_scale;
+
+    FixedOrScalableVector(const Type &type, bool scalable = true)
+        : type_pattern(type.is_vector() ? Fixed : Scalable),
+          type(type), relative_scale(1) {
+    }
+    FixedOrScalableVector(halide_type_code_t code)
+        : type_pattern(WildcardWidths),
+          type(code, 8, 1), relative_scale(1) {
+    }
+    FixedOrScalableVector(halide_type_code_t code, int relative_scale)
+        : type_pattern(WildcardWidths),
+          type(code, 8, 1), relative_scale(relative_scale) {
+    }
+    FixedOrScalableVector()
+        : type_pattern(Undefined),
+          type() {
+    }
+};
+
+struct RISCVIntrinsic {
+    const char *riscv_name;
+    FixedOrScalableVector ret_type;
+    const char *name;
+    FixedOrScalableVector arg_types[max_intrinsic_args];
+    int flags;
+    enum {
+        AddVLArg = 1 << 0,          // Add a constant full size vector length argument
+        RoundDown = 1 << 1,         // Set rounding mode to down (rdn) before intrinsic.
+        RoundUp = 1 << 2,           // Set rounding mode to up (rdu) before intrinsic.
```

Comment:
Could you provide a link in the comments that points to documentation of the rounding modes?

Reply:
Not in inline comments as it's pretty long and to a PDF so I can't (easily) link directly to the section anyway. Strikes me as the sort of thing where it's pretty obvious what they do and one has to know a bit about the RISC V vector architecture to work on the code at all. I added "vxrm" to make it clear these are the fixed-point rounding mode.